flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/data/StructRowData.java

```diff
@@ -0,0 +1,340 @@
+    }
+  }
+
+  @Override
+  public float getFloat(int pos) {
+    return struct.get(pos, Float.class);
+  }
+
+  @Override
+  public double getDouble(int pos) {
+    return struct.get(pos, Double.class);
+  }
+
+  @Override
+  public StringData getString(int pos) {
+    return isNullAt(pos) ? null : getStringDataInternal(pos);
+  }
+
+  private StringData getStringDataInternal(int pos) {
+    CharSequence seq = struct.get(pos, CharSequence.class);
+    return StringData.fromString(seq.toString());
+  }
+
+  @Override
+  public DecimalData getDecimal(int pos, int precision, int scale) {
+    return isNullAt(pos)
+        ? null
+        : DecimalData.fromBigDecimal(getDecimalInternal(pos), precision, scale);
+  }
+
+  private BigDecimal getDecimalInternal(int pos) {
+    return struct.get(pos, BigDecimal.class);
+  }
+
+  @Override
+  public TimestampData getTimestamp(int pos, int precision) {
+    long timeLong = getLong(pos);
+    if (precision == 6) {
+      int nanosOfMillisecond = (int) (timeLong % 1000);
+      return TimestampData.fromEpochMillis(timeLong / 1000, nanosOfMillisecond);
+    } else {
+      return TimestampData.fromEpochMillis(timeLong);
+    }
+  }
+
+  @Override
+  public <T> RawValueData<T> getRawValue(int pos) {
+    return RawValueData.fromBytes(getBinary(pos));
+  }
+
+  @Override
+  public byte[] getBinary(int pos) {
+    return isNullAt(pos) ? null : getBinaryInternal(pos);
+  }
+
+  private byte[] getBinaryInternal(int pos) {
+    Object bytes = struct.get(pos, Object.class);
+
+    // should only be either ByteBuffer or byte[]
+    if (bytes instanceof ByteBuffer) {
+      return ByteBuffers.toByteArray((ByteBuffer) bytes);
+    } else if (bytes instanceof byte[]) {
+      return (byte[]) bytes;
+    } else {
+      throw new IllegalStateException(
+          "Unknown type for binary field. Type name: " + bytes.getClass().getName());
+    }
+  }
+
+  @Override
+  public ArrayData getArray(int pos) {
+    return isNullAt(pos) ? null : getArrayInternal(pos);
+  }
+
+  private ArrayData getArrayInternal(int pos) {
+    return collectionToArrayData(
+        type.fields().get(pos).type().asListType().elementType(),
+        struct.get(pos, Collection.class));
+  }
+
+  @Override
+  public MapData getMap(int pos) {
+    return isNullAt(pos) ? null : getMapInternal(pos);
+  }
+
+  private MapData getMapInternal(int pos) {
+    return mapToMapData(type.fields().get(pos).type().asMapType(), struct.get(pos, Map.class));
+  }
+
+  @Override
+  public RowData getRow(int pos, int numFields) {
+    return isNullAt(pos) ? null : getStructRowData(pos, numFields);
+  }
+
+  private StructRowData getStructRowData(int ordinal, int numFields) {
+    return new StructRowData(
+        type.fields().get(ordinal).type().asStructType(), struct.get(ordinal, StructLike.class));
+  }
+
+  private MapData mapToMapData(Types.MapType mapType, Map<?, ?> map) {
+    // make a defensive copy to ensure entries do not change
```

Comment:
StructRowData just apply the view of RowData on an Iceberg Struct. not sure if we should worry about the underline struct changes or not.

Reply:
This ensures that it does not change when the  sub-query has a complex transformation, such as `select map['key']+ 1, map['key']+2`. Please correct me if I am mistaken.